net/HttpRequest.hpp

```diff
@@ -1621,11 +1621,13 @@ class Session final : public ProtocolHandlerInterface
     {
         _socket.reset(); // Reset to make sure we are disconnected.
 
-        auto callback = [this, &poll](std::shared_ptr<StreamSocket> socket) {
-            asyncConnectCompleted(poll, socket);
+        auto pushConnectCompleteToPoll = [this, &poll](std::shared_ptr<StreamSocket> socket) {
+            poll.addCallback([selfLifecycle = shared_from_this(), this, &poll, socket=std::move(socket)]() {
```

Comment:
I assume this is because we don't want races between the poll thread and the asyncDNS thread. A comment would be a helpful reminder in that case, I think.

Nit: Also, since we already have `selfLifecycle`, it might be simpler to use that instead of capturing (and using) `this`.

Reply:
wrt selfLifeCycle, it's a std::shared_ptr<ProtocolHandlerInterface> which doesn't have asyncConnectCompleted, so it turns into something like

std::static_pointer_cast<Session>(selfLifecycle)->asyncConnectCompleted(poll, socket);

which a little unfriendly to read, so I'm inclined towards the suggestion of https://devblogs.microsoft.com/oldnewthing/20190104-00/?p=100635 to "capture both a strong reference and a raw this. The strong reference keeps the this alive, and you use the this for convenient access to members."